sql/core/src/main/scala/org/apache/spark/sql/execution/GeneratedAggregate.scala

```diff
@@ -266,7 +266,18 @@ case class GeneratedAggregate(
 
       val joinedRow = new JoinedRow3
 
-      if (groupingExpressions.isEmpty) {
+      if (!iter.hasNext) {
+        // This is an empty input, so return early so that we do not allocate data structures
+        // that won't be cleaned up (see SPARK-8357).
+        if (groupingExpressions.isEmpty) {
```

Comment:
Here, I made a slight simplification compared to @navis's original patch: if `groupingExpressions` is empty and the input is empty, then always return an empty aggregation buffer.  @navis's patch contained an additional branch here which would skip this output if `partial = true`, but I think that is an unnecessary performance optimization given that the non-generated-Aggregate operator still outputs an empty row even on empty inputs.  Removing this branch means fewer cases to have to test.